src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1227,25 +1275,29 @@ private void parseLegacyPut(String name, String value) {
      */
     public Service getService(String type, String algorithm) {
         checkInitialized();
-
         // avoid allocating a new ServiceKey object if possible
         ServiceKey key = previousKey;
         if (key.matches(type, algorithm) == false) {
             key = new ServiceKey(type, algorithm, false);
             previousKey = key;
         }
+
         if (!serviceMap.isEmpty()) {
             Service s = serviceMap.get(key);
             if (s != null) {
                 return s;
             }
         }
-        synchronized (this) {
-            ensureLegacyParsed();
-            if (legacyMap != null && !legacyMap.isEmpty()) {
-                return legacyMap.get(key);
+
+        if (!legacyMap.isEmpty()) {
+            Service s = legacyMap.get(key);
+            if (s != null && !s.isValid()) {
+                legacyMap.remove(key);
```

Comment:
I wonder if it's better to use `legacyMap.remove(key, value);` here. What if another thread put some other value by the same key?

Reply:
Sure, I can do that.